# script for running Deseq2. Saved into "UVM/ecological_genomics_23/transcriptomics/scripts" folder
# created 10/18/23, alongside Transcriptomics Day 4
# set working directory
setwd("~/Documents/UVM/ecological_genomics_23/transcriptomics/results")

library(DESeq2)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(wesanderson)
library(vsn) # package within BiocManager
library(hexbin) # package within BiocManager
# BiocManager::install("DESeq2")
# First: BiocManager::install("vsn") AND BiocManager::install("hexbin")

#######################################

# Step 1: load in data! 

######################################

# Import the counts matrix
countsTable <- read.table("salmon.isoform.counts.matrix", header=TRUE, row.names=1)
head(countsTable)
dim(countsTable)

countsTableRound <- round(countsTable) # bc DESeq2 doesn't like decimals (and Salmon outputs data with decimals)
head(countsTableRound)

#import the sample discription table. Conditions
conds <- read.delim("ahud_samples_R.txt", header=TRUE, stringsAsFactors = TRUE, row.names=1)
head(conds)

#######################################

# Step 2: explore data distributions 

######################################

# Let's see how many reads we have from each sample
colSums(countsTableRound)
mean(colSums(countsTableRound))

barplot(colSums(countsTableRound), names.arg=colnames(countsTableRound),cex.names=0.5, las=3,ylim=c(0,21000000))
abline(h=mean(colSums(countsTableRound)), col="blue", lwd=2)

# the average number of counts per gene
rowSums(countsTableRound)
mean(rowSums(countsTableRound)) # [1] 2245.401
median(rowSums(countsTableRound)) # [1] 117

apply(countsTableRound,2,mean) # 2 in the apply function does the action across columns
apply(countsTableRound,1,mean) # 1 in the apply function does the action across rows
hist(apply(countsTableRound,1,mean),xlim=c(0,1000), ylim=c(0,300000),breaks=10000)

#######################################

# Step 3: start working with Deseq2

######################################
#### Create a DESeq object and define the experimental design here with the tilda

dds <- DESeqDataSetFromMatrix(countData = countsTableRound, colData=conds, 
                              design= ~ treatment + generation)

dim(dds) # 349516 x 38

# Filter out genes with too few reads - remove all genes with counts < 15 in more than 75% of samples, so ~28)
## suggested by WGCNA on RNAseq FAQ

dds <- dds[rowSums(counts(dds) >= 30) >= 28,] # 30 determined by looking at histogram
nrow(dds) # 41348
# choices are being made based on histogram. Getting rid of data that isn't very robust

# Run the DESeq model to test for differential gene expression
dds <- DESeq(dds)

# List the results you've generated
resultsNames(dds)
# compares across generation pairs and treatments
# didn't test for an interaction between factors